drivers/soundwire/intel_init.c

```diff
@@ -257,24 +321,17 @@ static struct sdw_intel_ctx
 		link->shim_mask = &ctx->shim_mask;
 		link->link_mask = link_mask;
 
-		memset(&pdevinfo, 0, sizeof(pdevinfo));
-
-		pdevinfo.parent = res->parent;
-		pdevinfo.name = "intel-sdw";
-		pdevinfo.id = i;
-		pdevinfo.fwnode = acpi_fwnode_handle(adev);
-		pdevinfo.data = link;
-		pdevinfo.size_data = sizeof(*link);
-
-		pdev = platform_device_register_full(&pdevinfo);
-		if (IS_ERR(pdev)) {
-			dev_err(&adev->dev,
-				"platform device creation failed: %ld\n",
-				PTR_ERR(pdev));
-			goto err;
+		ldev->link = link;
+
+		ret = ancillary_device_add(&ldev->ancildev);
+		if (ret < 0) {
+			dev_err(&adev->dev, "failed to add link dev %s\n", ldev->ancildev.name);
+			put_device(&ldev->ancildev.dev);
+			goto err; /* FIXME error flow */
 		}
-		link->pdev = pdev;
-		link->cdns = platform_get_drvdata(pdev);
+
+		link->ldev = ldev;
```

Comment:
this is what I find a bit odd. setting ldev->link = link  and link->ldev = ldev.

Reply:
such cross-linking does feel odd and you usually try to avoid it, but sometimes it's so damn useful :-) And yes, there are such examples around already, I think